include/boost/geometry/algorithms/detail/overlay/split_rings.hpp

```diff
@@ -0,0 +1,433 @@
+            >(ring, robust_policy, turns, policy);
+    }
+
+    static inline void get_self_turns(Ring const& ring,
+                                      turns_container_type& turns,
+                                      RobustPolicy const& robust_policy)
+    {
+        get_self_turns(ring, turns, robust_policy, no_interrupt_policy());
+    }
+
+    template <typename MUU_Turns, typename RingOut>
+    static inline void insert_muu_turns(Ring const& ring,
+                                        MUU_Turns const& muu_turns,
+                                        RingOut& ring_out)
+    {
+        typedef typename boost::range_iterator
+            <
+                MUU_Turns const
+            >::type iterator_type;
+
+        typename boost::range_size<Ring>::type point_index = 0;
+        for (iterator_type it = muu_turns.begin(); it != muu_turns.end(); ++it)
+        {
+            signed_size_type segment_index
+                = get_correct_op(*it).seg_id.segment_index;
+
+            while (point_index <= segment_index)
+            {
+                ring_out.push_back(ring[point_index]);
+                ++point_index;
+            }
+
+            ring_out.push_back(it->point);
+        }
+
+        while (point_index < ring.size())
+        {
+            ring_out.push_back(ring[point_index]);
+            ++point_index; 
+        }
+    }
+
+    template <typename RingIn, typename Collection>
+    static inline void copy_to_collection(RingIn const& ring,
+                                          Collection& collection)
+    {
+        typedef typename boost::range_value<Collection>::type ring_out_type;
+
+        ring_out_type tmp;
+        for (typename boost::range_iterator<RingIn const>::type
+                 it = ring.begin();
+             it != ring.end();
+             ++it)
+        {
+            geometry::traits::push_back<ring_out_type>::apply(tmp, *it);
+        }
+        collection.push_back(tmp);
+    }
+
+    template <closure_selector Closure, typename RingType, typename Collection>
+    static inline void split_one_ring(RingType& ring, Collection& collection)
+    {
+        // create and initialize a stack with the input ring
+        std::stack<RingType> stack;
+        stack.push(RingType());
+        stack.top().swap(ring);
+
+        // while the stack is not empty:
+        // look for duplicates, split and push to stack;
+        // otherwise, copy to output collection
+        while (! stack.empty())
+        {
+            RingType& top_ring = stack.top();
+
+            typename boost::range_iterator<RingType>::type pos1, pos2;
+            bool duplicate_found = find_duplicate_points
+                <
+                    Closure
+                >::apply(top_ring, pos1, pos2);
+
+            if (duplicate_found)
+            {
+                RingType other_ring;
+                stack.push(other_ring);
+                top_ring.split_at(pos1, pos2, other_ring);
```

Comment:
`top_ring` is used after `stack.push()` this may cause SEGFAULT since during push the stack may be resized so all elements copied into a different location. The `top_ring` should be split first and then the result pushed or moved using this function I mentioned in a different commit. The algorithm would also be more clear this way.